src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -809,6 +814,33 @@ private static long CountOfRevisions(DocumentsOperationContext context, Slice pr
             return numbers.Read(prefix)?.Reader.ReadLittleEndianInt64() ?? 0;
         }
 
+        public Document GetRevisionBefore(DocumentsOperationContext context, string id, DateTime max)
+        {
+            using (DocumentIdWorker.GetSliceFromId(context, id, out Slice lowerId))
+            using (GetKeyPrefix(context, lowerId, out Slice prefixSlice))
+            using (GetLastKey(context, lowerId, out Slice lastKey))
+            {
+                // Here we assume a reasonable number of revisions and scan the entire history
+                // This is because we want to handle out of order revisions from multiple nodes so the local etag
+                // order is different than the last modified order
+                Document result = null;
+                var table = new Table(RevisionsSchema, context.Transaction.InnerTransaction);
+                foreach (var tvr in table.SeekBackwardFrom(RevisionsSchema.Indexes[IdAndEtagSlice], prefixSlice, lastKey, 0))
+                {
+                    var document = TableValueToRevision(context, ref tvr.Result.Reader);
+                    if (document.LastModified > max)
+                        continue;
+                    if(result == null || 
```

Comment:
Can we assume here that the revisions are ordered by data here?
Because if we have hundreds of revisions, but the most relevant are the first 50, then we iterate over 150 unneeded revisions..
Also, maybe we could avoid caliing TableValueToRevision for each revision, and instead call to a function that will return the raw dateTime value